### PR TITLE
[Misc] [refactor] Move symbol versioning to a new file

### PR DESCRIPTION
**Pull request by qiao-bo**
Create a new `symbol_version.cpp` for pinning math symbols to older GLIBC versions. 

This PR also documents the method to pin offending math symbols back to GLIBC 2.2.5:
Related issue = #3174 
Reference solution: https://github.com/taichi-dev/taichi/pull/1342

Taichi build on a Ubuntu 20.04
Before:
```
$ objdump -T libtaichi_core.so | grep GLIBC_2.29     
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.29  exp                                                      
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.29  log                                                      
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.29  pow                                                      
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.29  log2    
```
After: `objdump -T libtaichi_core.so | grep GLIBC_2.29 ` gives none. current glibc versioning is
```
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_join
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3.4 __snprintf_chk
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 ftello64
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 dlsym
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strdup  
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 feof  
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 uname    
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 __lxstat64  
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 symlink  
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 realloc
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 access
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 isalpha  
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 memmove              
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3   __ctype_b_loc
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strncpy
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 toupper
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 localtime_r   
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 __assert_fail
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 mallinfo
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 remove               
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 syscall 
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_setspecific
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 __lxstat 
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 unlink          
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3.4 __fprintf_chk
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 ftell    
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 gmtime  
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 mkdir  
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 lseek                           
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 atan2         
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 readlink   
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.4   __stack_chk_fail                
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 memset 
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strtok_r        
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 dlclose
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 getuid               
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fputs 
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 sin                                             
0000000000000000  w   DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_once                                             
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 mktime                                                   
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strlen
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fclose
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3   __ctype_tolower_loc
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 isalnum      
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strstr
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 statfs         
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 sqrt              
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fflush            
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_key_delete
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3.4 __memcpy_chk
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 acosf 
0000000000000000      DO *UND*  0000000000000000  GLIBC_2.2.5 stdout             
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 posix_spawn_file_actions_init
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 opendir
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3   realpath         
0000000000000000      DO *UND*  0000000000000000  GLIBC_2.2.5 stderr 
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 sigfillset    
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 powf        
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 sigaltstack
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 freopen64
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 vsnprintf
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 islower
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 sinh 
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_self
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 utime
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 posix_spawn_file_actions_adddup2
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 isspace  
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strtoll
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 sigaction
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 gmtime_r
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_getspecific
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 popen    
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 sigemptyset
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 backtrace_symbols_fd
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strtol
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_attr_setstacksize
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 __fxstat64
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 __cxa_atexit
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 round   
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 isascii             
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 isatty
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 execv 
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 waitpid
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 cos      
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 sigaddset
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 lseek64
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 abs        
0000000000000000  w   DF *UND*  0000000000000000  GLIBC_2.2.5 __pthread_key_create
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 modf       
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 frexpf 
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 feclearexcept
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 umask
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 getppid
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 sprintf              
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 nextafterf
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 ldexpf 
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 usleep         
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 memcmp
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fwrite
0000000000000000      DO *UND*  0000000000000000  GLIBC_2.2.5 environ 
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 log  
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 __xstat 
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fileno
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fseek    
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 atan        
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 backtrace
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fopen
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 system
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 ftruncate
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_setschedparam
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 wait
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 sinf
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 dlerror
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fseeko64      
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 printf
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.14  memcpy
0000000000000000  w   DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_rwlock_wrlock
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 mprotect
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 _setjmp
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 nextafter
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 chmod           
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 free    
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pow      
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.6   futimens
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fprintf
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.12  pthread_getname_np              
0000000000000000  w   DF *UND*  0000000000000000  GLIBC_2.2.5 __cxa_finalize
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strtof     
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_attr_init               
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 getchar
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 getpwuid        
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 close
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3.4 __vsnprintf_chk      
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 posix_fallocate
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strncmp           
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 __xmknod
0000000000000000  w   DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_mutex_unlock
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 cosh
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 alarm
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 log2f
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.11  __longjmp_chk
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 perror
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 dl_iterate_phdr
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_key_create
0000000000000000  w   DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_mutex_lock
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3.4 sched_getaffinity
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 signal
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 getcwd
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3   __ctype_toupper_loc
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 memchr
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 open
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 backtrace_symbols
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 isdigit
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.3   __tls_get_addr
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 gettimeofday
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 log10
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strchr
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 snprintf
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 tan
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fread
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 sysconf
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 dup2
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 closedir
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strsignal
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 munmap
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 acos
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fchmod
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 localtime
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 __xstat64
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 exit
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 wcslen
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 cosf
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_create
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 log2
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 time
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 getpwnam
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_attr_destroy
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 qsort
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 getenv
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 abort
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 setrlimit
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 read
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 readdir
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.15  posix_spawn
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 atof
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 sigprocmask
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fstatfs
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strtoull
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 asin
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strcmp
0000000000000000  w   DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_rwlock_rdlock
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 dlopen
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 tolower
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_sigmask
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 calloc
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 kill
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 isxdigit
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 raise
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strftime
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 link
0000000000000000      DO *UND*  0000000000000000  GLIBC_2.2.5 __environ
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fetestexcept
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 malloc
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 tanh
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 mmap
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strrchr
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 exp
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 rand
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fork
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 _exit
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.4   __realpath_chk
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 write
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 sqrtf
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 getrlimit
0000000000000000  w   DO *UND*  0000000000000000  GLIBC_2.2.5 _environ
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 execve
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 fopen64
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 __errno_location
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 __fxstat
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 getrusage
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 rename
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 getpid
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 posix_spawn_file_actions_destroy
0000000000000000      DO *UND*  0000000000000000  GLIBC_2.2.5 stdin
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 getpagesize
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 posix_spawn_file_actions_addopen
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 srand
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.6   __sched_cpucount
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 pread
0000000000000000  w   DF *UND*  0000000000000000  GLIBC_2.2.5 pthread_rwlock_unlock
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 chdir
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.12  pthread_setname_np
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strtod
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.2.5 strerror_r

```

### CODE REVIEW

File: taichi/common/core.cpp [hunk 1]
```diff
@@ -26,6 +26,22 @@ __asm__(".symver log2f,log2f@GLIBC_2.2.5");
 float __wrap_log2f(float x) {
```
Comment (k-ye) [inline taichi/common/core.cpp:26]: nit: Let's move these wrappers to a separate file to hightlight their importance

Comment (qiao-bo) [reply taichi/common/core.cpp:26]: how about to a `symbol_version.h`?

Comment (k-ye) [reply taichi/common/core.cpp:26]: nit: `symbol_version.cpp` (IIUC, we only need these function to be defined)?

Comment (qiao-bo) [reply taichi/common/core.cpp:26]: > Can we also add link to the issue in both files? ;)

you mean link to this issue? #3174

**Comment by qiao-bo** [comment]
> Can we also add link to the issue in both files? ;)

you mean link to this issue? #3174

**Comment by ailzhang** [comment]
@qiao-bo Yup either issue or PR or a short note describing the problem it's trying to solve will help a lot! Thanks!

**Review by k-ye — APPROVED** [review]
Great job, LGTM!

**Comment by qiao-bo** [comment]
@k-ye @ailzhang, The GLIBC 2.2.5 still cannot be used at the moment due to Vulkan. Nevertheless, this PR also moves the symbol version code to a new file and i think this part can be merged. I renamed it and opened a new issue #3552 to track next steps when we have time in the future. what do you think?

**Review by k-ye — APPROVED** [review]
LGTM!

**Pull request merged**